go-controller/pkg/ovn/controller/services/services_controller.go

```diff
@@ -207,6 +208,47 @@ func (c *Controller) Run(workers int, stopCh <-chan struct{}, runRepair, useLBGr
 		c.repair.runBeforeSync(c.useTemplates)
 	}
 
+	// Ensure service node cache is up-to-date before we allow service sync to start.
+	// This avoids a case where ovnkube-controller may be restarting, the node cache is not up-to-date,
+	// a service is rendered with a node missing, and its per node LB is incorrectly removed.
+	maxTimeout := 10 * time.Second
+	klog.Infof("Waiting up to %s for service node cache to be synced", maxTimeout)
+	start := time.Now()
+	var err1 error
+	err = wait.PollUntilContextTimeout(ctx, 250*time.Millisecond, maxTimeout, true, func(ctx context.Context) (bool, error) {
```

Comment:
@trozet I'm struggling to think of why HasSynced() isn't doing all the work here, and I think see a couple issues.

1. When the node tracker starts it attaches an Event Handler to the shared informer, and that handler will be delivered the initial set of nodes 1-by-1. It calls updateNode() which eventually calls updateNodeInfo() which then calls RequestFullSync(). So every time the node tracker gets one more node from the apiserver in it's initial event handler attachment, we'll be sending the service controller a partial node list. RequestFullSync() will then call onServiceAdd() for each service, but that'll end up in syncService() with a partial node list.

Couldn't there be a way to have the node tracker cache the returned event handler in Start() and then check
HasSynced() in updateNodeInfo()? If it's false then don't call nt.resyncFn() yet. The only missing bit is how you know when syncing is over so you *can* call nt.resyncFn(); the dumb implementation would just assume that some node event is going to come shortly after syncing is finished.  Maybe that's where the 10s timer could live, just force a `nt.resyncFn(nt.getZoneNodes())` 10s afer nodeTracker attaches its' handler?

2. Along with that, in the Service Controller's Run() all the handlers' get attached, but we're *not waiting* for their initial syncs to finish, we're just waiting for the *parent* informer to finish. We should be doing something like this:

```
	serviceHandler, err := c.serviceInformer.Informer().AddEventHandler(factory.WithUpdateHandlingForObjReplace(cache.ResourceEventHandlerFuncs{
		AddFunc:    c.onServiceAdd,
		UpdateFunc: c.onServiceUpdate,
		DeleteFunc: c.onServiceDelete,
	}))
	if err != nil {
		return err
	}

	klog.Info("Setting up event handlers for endpoint slices")
	epHandler, err = c.endpointSliceInformer.Informer().AddEventHandler(factory.WithUpdateHandlingForObjReplace(cache.ResourceEventHandlerFuncs{
		AddFunc:    c.onEndpointSliceAdd,
		UpdateFunc: c.onEndpointSliceUpdate,
		DeleteFunc: c.onEndpointSliceDelete,
	}))
	if err != nil {
		return err
	}

	nodeHasSynced, err = c.nodeTracker.Start(c.nodeInformer)
	if err != nil {
		return err
	}

	// Wait for the caches to be synced
	klog.Info("Waiting for informer caches to sync")
	if !util.WaitForNamedCacheSyncWithTimeout(controllerName, stopCh, serviceHandler.HasSynced, epHandler.HasSynced, nodeHasSynced) {
		return fmt.Errorf("error syncing cache")
	}
```

Because the returned handlers are actually an interface with a HasSynced() function that checks both whether the parent has synced *and* whether all the initial adds have been processed.

Right now we're running all the Service, EP, and Node processing in parallel at startup but we're not even waiting in Run() for them all to finish, I think.

Reply:
actually this is a good point, don't we like wait and sync in service controller to get all the nodes in the cache....? we don't start till then right ? (/me confused)

Reply:
Ah I didn't realize the HasSynced function worked like that. I thought it was only for the informer and not initial add. The 10 seconds really shouldn't be needed if we just force nodeTracker to finish its initial add before the service or node handlers are started. Let me try fixing it that way and see what happens.

Reply:
@trozet @tssurya HasSynced() works differently for the informer and the event handler. For the informer it waits for the cache to be populated, but for the informer it waits for it's parent Informer to sync *and* it waits for the initial add to happen.

This didn't used to be the case, the event handler return got added a couple Kube releases ago so it's pretty new.

Aside: you can now remove handlers from the informer using that returned handle, which means we could theoretically get rid of most of the factory.go code since that was it's only reason in life: to be able to remove handlers, which shared informer didn't have until last year some time.